programs/drift/src/state/fulfillment_params/phoenix.rs

```diff
@@ -0,0 +1,479 @@
+        let quote_mint = self.phoenix_market.header.quote_params.mint_key;
+
+        let new_order_instruction = create_new_order_instruction_with_custom_token_accounts(
+            self.phoenix_market.key,
+            self.drift_signer.key,
+            &self.base_market_vault.key(),
+            &self.quote_market_vault.key(),
+            &base_mint,
+            &quote_mint,
+            &order_packet,
+        );
+
+        let signer_seeds = get_signer_seeds(&self.signer_nonce);
+        let signers_seeds = &[&signer_seeds[..]];
+
+        invoke_signed_unchecked(
+            &new_order_instruction,
+            &self.to_account_infos(),
+            signers_seeds,
+        )
+        .map_err(|e| {
+            msg!("{:?}", e);
+            ErrorCode::FailedPhoenixCPI
+        })?;
+
+        Ok(())
+    }
+}
+
+impl<'a, 'b> SpotFulfillmentParams for PhoenixFulfillmentParams<'a, 'b> {
+    fn is_external(&self) -> bool {
+        true
+    }
+
+    fn get_best_bid_and_ask(&self) -> DriftResult<(Option<u64>, Option<u64>)> {
+        // Conversion: price_in_ticks (T) * tick_size (QL/BU * T) * quote_lot_size (QA/QL) / raw_base_units_per_base_unit (rBU/BU)
+        // Yields: price (QA/rBU)
+        let market_data = self.phoenix_market.data.borrow();
+        let (_, market_bytes) = market_data.split_at(size_of::<MarketHeader>());
+        let header = &self.phoenix_market.header;
+        if header.quote_params.decimals != 6 {
+            msg!("Quote decimals must be 6");
+            return Err(ErrorCode::InvalidPricePrecision);
+        }
+
+        let market = load_with_dispatch(&header.market_size_params, market_bytes)
+            .map_err(|_| {
+                msg!("Failed to deserialize market");
+                ErrorCode::FailedToDeserializePhoenixMarket
+            })?
+            .inner;
+
+        let best_bid = market.get_book(Side::Bid).iter().next().and_then(|(o, _)| {
+            o.price_in_ticks
+                .as_u64()
+                .safe_mul(market.get_tick_size().as_u64())
+                .ok()?
+                .safe_mul(header.get_quote_lot_size().as_u64())
+                .ok()?
+                .safe_div(header.raw_base_units_per_base_unit as u64)
+                .ok()
+        });
+        let best_ask = market.get_book(Side::Ask).iter().next().and_then(|(o, _)| {
+            o.price_in_ticks
+                .as_u64()
+                .safe_mul(market.get_tick_size().as_u64())
+                .ok()?
+                .safe_mul(header.get_quote_lot_size().as_u64())
+                .ok()?
+                .safe_div(header.raw_base_units_per_base_unit as u64)
+                .ok()
+        });
+        Ok((best_bid, best_ask))
+    }
+
+    fn fulfill_order(
+        &mut self,
+        taker_direction: PositionDirection,
+        taker_price: u64,
+        taker_base_asset_amount: u64,
+        _taker_max_quote_asset_amount: u64,
```

Comment:
since phoenix doesn't support max quote, should account for this by making the base amount equal to the min of the `taker_base_asset_amount` and `taker_max_quote_asset_amount / price`